admission-policy/readme.adoc

```diff
@@ -0,0 +1,405 @@
+```bash
+[req]
+req_extensions = v3_req
+distinguished_name = req_distinguished_name
+[req_distinguished_name]
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+extendedKeyUsage = clientAuth, serverAuth
+subjectAltName = @alt_names
+[alt_names]
+DNS.1 = opa.opa.cluster.svc.local
+```
+
+*server.conf*
+```bash
+[req]
+req_extensions = v3_req
+distinguished_name = req_distinguished_name
+[req_distinguished_name]
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+extendedKeyUsage = clientAuth, serverAuth
+subjectAltName = @alt_names
+[alt_names]
+DNS.1 = opa.opa.cluster.svc.local
+```
+
+IMPORTANT: The subjectAltName/IP address in the certificate MUST match the one configured on the Kubernetes Service.
+
+Now create local files that contain the CA and client/server key pairs.  Shortly we will
+hand these keys to OPA and the API server so they can communicate.
+
+Create a certificate authority:
+
+```bash
+openssl genrsa -out ca.key 2048
+```
+
+```bash
+openssl req -x509 -new -nodes -key ca.key -days 100000 -out ca.crt -subj "/CN=admission_ca"
+```
+
+Create a server certiticate:
+
+```bash
+openssl genrsa -out server.key 2048
+```
+
+```bash
+openssl req -new -key server.key -out server.csr -subj "/CN=admission_server" -config server.conf
+```
+
+```bash
+openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 100000 -extensions v3_req -extfile server.conf
+```
+
+Create a client certiticate:
+
+```bash
+openssl genrsa -out client.key 2048
+```
+
+```bash
+openssl req -new -key client.key -out client.csr -subj "/CN=admission_client" -config client.conf
+```
+
+```bash
+openssl x509 -req -in client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client.crt -days 100000 -extensions v3_req -extfile client.conf
+```
+
+
+=== 2.2: Deploy OPA on Kubernetes
+
+First, create a namespace to deploy OPA into.
+
+```bash
+kubectl create namespace opa
+```
+
+Create a Service to expose the OPA API. The Kubernetes API server will lookup
+the Service and execute webhook requests against it.
+
+**opa-admission-controller-service.yaml**:
+
+```yaml
+kind: Service
+apiVersion: v1
+metadata:
+  name: opa
+spec:
+  selector:
+    app: opa
+  ports:
+  - name: https
+    protocol: TCP
+    port: 443
+    targetPort: 443
+```
+
+```bash
+kubectl create -f opa-admission-controller-service.yaml -n opa
+```
+
+Next, create Secrets containing the TLS credentials for OPA:
+
+```bash
+kubectl create secret generic opa-ca --from-file=ca.crt -n opa
+kubectl create secret tls opa-server --cert=server.crt --key=server.key -n opa
+```
+
+Finally, create the Deployment to run OPA as an Admission Controller.
+The deployment contains two containers: `opa` and `kube-mgmt`.  `opa` by
+itself is a general-purpose policy engine and knows nothing about Kubernetes.
+kube-mgmt is a collection of Kubernetes-specific code that helps
```

Comment:
we should wrap kube-mgmt in backticks like `opa`

Reply:
Done.  See next patch.